README.md
jwi - Java Wordnet Interface
============================

(Mavenized version of) JWI 2.3.3 - this git repo is intended to shadow 
the latest source version from http://projects.csail.mit.edu/jwi/, but
augment the code with a maven *pom.xml* file with http://repo.delver.io
as the initial repository.

**Note:** No additional functionality (beyond mavenization) will _ever_ be added.

License
-------
See **jwi-license** file for details. All rights of original author reserved.


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/delver/jwi/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



changelog
/******************************************************************************
 * Copyright (c) 2007-2014 Massachusetts Institute of Technology.
 * All rights reserved.
 * 
 * This program and the accompanying materials are made available under the 
 * terms of the MIT Java Wordnet Interface License which accompanies this 
 * distribution.  Please check the attached license for more details.
 *
 * To communicate praise, suggestions, constructive criticism, or feature
 * requests, write:
 *
 * Mark Alan Finlayson
 * email: markaf@alum.mit.edu
 * post: 32 Vassar St., Room 32-258, Cambridge, MA 02138 USA
 * web: http://www.mit.edu/~markaf/
 *******************************************************************************/

Changelog for edu.mit.jwi:

Ver     : Release Date  : Changes
=================================
2.3.3   : Feb 12, 2014  : 1. Fixed Bug 023: RAMDictionary now properly handles underspecified word id objects
----------------------------------
2.3.2   : Dec 26, 2013  : 1. Fixed Bug 022: Minor changes to stemmer to correct lemma order and correctly handle nouns and verbs of two characters or fewer
----------------------------------
2.3.1   : Dec 09, 2013  : 1. Fixed Bug 021: Factored out IStemmingRule, and added check for nouns that end with 'ss'
----------------------------------
2.3.0   : Aug 28, 2013  : 1. Added IIndexWord.getPointers() method and IndexWord implementation thereof
                          2. Added two IndexWord constructors allowing setting of pointers field
                          3. Added IndexWordParser.resolvePointer(String, POS) method and its use in the parseLine(String) method
                          4. Added Pointer.DOMAIN (symbol ';') and Pointer.MEMBER (symbol '-') so IndexWordParser.parseLine(String) doesn't throw an exception
----------------------------------
2.2.6   : Jun 25, 2013  : 1. Fixed Bug 020
---------------------------------- 
2.2.5   : Jun 17, 2013  : 1. Fixed Bug 019
                          2. IItemID now implements Serializable 
---------------------------------- 
2.2.4   : Apr 13, 2013  : 1. Fixed Bug 017; added relevant FAQ
                          2. Fixed Bug 018; tested on 3.0 database-only version; added relevant FAQ
                          3. Tested JWI on Wordnet 3.1; added Version.ver31 field 
----------------------------------        
2.2.3   : Feb 25, 2012  : 1. Fixed Bug 015
----------------------------------
2.2.2   : Nov 16, 2011  : 1. Fixed Bug 014
                          2. Changed license to CC-BY, allowing commercial use
----------------------------------
2.2.1   : Jun 15, 2011  : 1. Fixed Bug 013
----------------------------------
2.2.0   : May 17, 2011  : 1. Overhauled Javadoc and Manual
                          2. Implemented in-memory dictionary
                          3. Created IHasLifecycle and ICloseable interfaces, now implemented by IDictionary, IDataProvider, IDataSource, and ItemCache
                              a) IDataSource objects must now be opened and closed via the open() and close() methods
                              b) IDictionary.open() method now throws an IOException
                              c) Renamed DictionaryClosedException to ObjectClosedException, and moved it into the IHasLifecycle interface
                          4. Fixed Bug 012
                              a) IDataSource now implements IHasLifecycle
                              b) FileProvider now calls IDataSource.open() and .close() at appropriate places 
                          5. Condensed iterator methods in WordnetFile hierarchy
                              a) Moved iterator() and iterator(String) into WordnetFile class from subclasses
                              b) Added WordnetFile.makeIterator(ByteBuffer, String) abstract method and implemented in subclasses
                          6. Abstracted Dictionary class hierarchy a bit; moved caching functionality into one class (CachingDictionary), data source access into another (DataSourceDictionary)
                          7. Removed IDataSourceDictionary#setDataProvider and #getSources() methods, added #getTypes() method
                          8. Created the ILoadable and ILoadPolicy interfaces that allow data in WordnetFile objects to be loaded completely into memory
                          9. Changed a few method contracts, making null-handling more consistent
                          10. Renamed IWordID.getSenseNumber() to getWordNumber()
----------------------------------
2.1.5   : Dec 19, 2008  : 1. Added documentation that points from the ISenseKey interface to SenseKeyParser class
                          2. Fixed Bug 011 (WordnetStemmer.findStems(String, POS))
                          3. Make explicit that Dictionary class does no stemming, unlike Wordnet API
                          4. Iterator returned by IDictionary.getExceptionEntryIterator(POS) now returns IExceptionEntry objects
                          5. Added catch and warning for linefeed corrupted data files
---------------------------------
2.1.4   : Sep 11, 2008  : 1. Moved SenseKey.parseKey(String) to its own class, SenseKeyParser, for ease of modification
                          2. Replaced call to SenseKey.parseKey(String) in SenseLineParser with call to SenseKeyParser.parseLine(String)
                          3. Added UnknownLexFile class to allow support of Wordnet mods (such as Rion Snow's) that use non-standard lexical file numbers
                          4. Modified SenseKeyParser and DataLineParser so they use UnknownLexFile class rather than throw NullPointerException on non-standard lexical file
                          5. Fixed Bug 010 (Dictionary.getExceptionEntry(IExceptionEntryID) method)
                          6. Replaced Pattern.quote(String) in Version class with string concatenation with \Q and \E for compatibility with GCJ 4.2

src/edu/mit/jwi/item/WordID.java
/********************************************************************************
 * MIT Java Wordnet Interface Library (JWI) v2.3.3
 * Copyright (c) 2007-2014 Massachusetts Institute of Technology
 *
 * JWI is distributed under the terms of the Creative Commons Attribution 3.0 
 * Unported License, which means it may be freely used for all purposes, as long 
 * as proper acknowledgment is made.  See the license file included with this
 * distribution for more details.
 *******************************************************************************/

package edu.mit.jwi.item;

/**
 * Default implementation of the {@code IWordID} interface.
 * 
 * @author Mark A. Finlayson
 * @version 2.3.3
 * @since JWI 1.0
 */
public class WordID implements IWordID {
	
	/** 
	 * Generated serial version id.
	 *
	 * @since JWI 2.2.5
	 */
	private static final long serialVersionUID = 3163309710173885763L;

	/** 
	 * String prefix for the {@link #toString()} method.
	 *
	 * @since JWI 2.0.0
	 */
	public static final String wordIDPrefix = "WID-";
	
	/** 
	 * Represents an unknown lemma for the {@link #toString()} method.
	 *
	 * @since JWI 2.0.0
	 */
	public static final String unknownLemma = "?";
	
	/** 
	 * Represents an unknown word number for the {@link #toString()} method.
	 *
	 * @since JWI 2.0.0
	 */
	public static final String unknownWordNumber = "??";
	
	// final instance fields
    private final ISynsetID id;
    private final String lemma;
    private final int num;

	/**
	 * Constructs a word id from the specified arguments. This constructor
	 * produces a word with an unknown lemma.
	 * 
	 * @param offset
	 *            the synset offset
	 * @param pos
	 *            the part of speech; may not be <code>null</code>
	 * @param num
	 *            the word number
	 * @throws IllegalArgumentException
	 *             if the offset or number are not legal
	 * @since JWI 1.0
	 */
    public WordID(int offset, POS pos, int num){
    	this(new SynsetID(offset, pos), num);
    }
    
	/**
	 * Constructs a word id from the specified arguments. This constructor
	 * produces a word with an unknown word number.
	 * 
	 * @param offset
	 *            the synset offset
	 * @param pos
	 *            the part of speech; may not be <code>null</code>
	 * @param lemma
	 *            the lemma; may not be <code>null</code>, empty, or all
	 *            whitespace
	 * @since JWI 1.0
	 */
    public WordID(int offset, POS pos, String lemma){
    	this(new SynsetID(offset, pos), lemma);
    }

	/**
	 * Constructs a word id from the specified arguments. This constructor
	 * produces a word with an unknown lemma.
	 * 
	 * @param id
	 *            the synset id; may not be <code>null</code>
	 * @param num
	 *            the word number
	 * @throws NullPointerException
	 *             if the synset id is <code>null</code>
	 * @throws IllegalArgumentException
	 *             if the lemma is empty or all whitespace
	 * @since JWI 1.0
	 */
    public WordID(ISynsetID id, int num) {
        if (id == null)
        	throw new NullPointerException();
        Word.checkWordNumber(num);
        this.id = id;
        this.num = num;
        this.lemma = null;
    }

	/**
	 * Constructs a word id from the specified arguments. This constructor
	 * produces a word with an unknown word number.
	 * 
	 * @param id
	 *            the synset id; may not be <code>null</code>
	 * @param lemma
	 *            the lemma; may not be <code>null</code>, empty, or all
	 *            whitespace
	 * @throws NullPointerException
	 *             if the synset id is <code>null</code>
	 * @throws IllegalArgumentException
	 *             if the lemma is empty or all whitespace
	 * @since JWI 1.0
	 */
    public WordID(ISynsetID id, String lemma) {
        if(id == null)
        	throw new NullPointerException();
        if(lemma.trim().length() == 0)
        	throw new IllegalArgumentException();
        this.id = id;
        this.num = -1;
        this.lemma = lemma.toLowerCase();
    }
    
	/**
	 * Constructs a fully specified word id
	 * 
	 * @param id
	 *            the synset id; may not be <code>null</code>
	 * @param num
	 *            the word number
	 * @param lemma
	 *            the lemma; may not be <code>null</code>, empty, or all
	 *            whitespace
	 * @throws NullPointerException
	 *             if the synset id is <code>null</code>
	 * @throws IllegalArgumentException
	 *             if the lemma is empty or all whitespace, or the word number
	 *             is not legal
	 * @since JWI 1.0
	 */
    public WordID(ISynsetID id, int num, String lemma){
    	if(id == null)
    		throw new NullPointerException();
    	if(lemma.trim().length() == 0)
    		throw new IllegalArgumentException();
    	Word.checkWordNumber(num);
        this.id = id;
        this.num = num;
        this.lemma = lemma;
    }

    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.item.IWordID#getSynsetID()
     */
    public ISynsetID getSynsetID() {
        return id;
    }

    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.item.IWordID#getWordNumber()
     */
    public int getWordNumber() {
        return num;
    }

    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.item.IWordID#getLemma()
     */
    public String getLemma() {
        return lemma;
    }

    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.item.IHasPOS#getPOS()
     */
    public POS getPOS() {
        return id.getPOS();
    }

    /* 
     * (non-Javadoc) 
     *
     * @see java.lang.Object#hashCode()
     */
    @Override
    public int hashCode() {
    	return 31*id.hashCode();
//        final int PRIME = 31;
//        int result = 1;
//        result = PRIME * result + ((id == null) ? 0 : id.hashCode());
//        result = PRIME * result + num;
//        if (lemma != null)
//        	result = PRIME * result + lemma.hashCode();
//        return result;
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#equals(java.lang.Object)
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj)
        	return true;
        if (obj == null)
        	return false;
        if (getClass() != obj.getClass())
        	return false;
        final WordID other = (WordID) obj;
        if (!id.equals(other.id))
        	return false;
        if (other.num != -1 && num != -1 && other.num != num)
        	return false;
        if (other.lemma != null && lemma != null)
            if (!other.lemma.equalsIgnoreCase(lemma))
            	return false;
        return true;
    }
    
    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#toString()
     */
    @Override
    public String toString() {
    	StringBuilder sb = new StringBuilder();
    	sb.append(wordIDPrefix);
    	sb.append(Synset.zeroFillOffset(id.getOffset()));
    	sb.append('-');
    	sb.append(Character.toUpperCase(id.getPOS().getTag()));
    	sb.append('-');
    	sb.append((num < 0) ? unknownWordNumber : Word.zeroFillWordNumber(num));
        sb.append('-');
        sb.append((lemma == null) ? unknownLemma : lemma);
        return sb.toString();
    }

	/**
	 * Parses the result of the {@link #toString()} method back into an
	 * {@code WordID}. Word id's are always of the following format:
	 * WID-########-P-##-lemma where ######## is the eight decimal digit
	 * zero-filled offset of the associated synset, P is the upper case
	 * character representing the part of speech, ## is the two hexidecimal
	 * digit zero-filled word number (or ?? if unknown), and lemma is the lemma.
	 * 
	 * @param value
	 *            the string to be parsed
	 * @return WordID the parsed id
	 * @throws IllegalArgumentException
	 *             if the specified string does not represent a word id
	 * @throws NullPointerException
	 *             if the specified string is <code>null</code>
	 * @since JWI 1.0
	 */
    public static IWordID parseWordID(String value) {
    	if(value == null)
    		throw new NullPointerException();
        if (value.length() < 19)
        	throw new IllegalArgumentException();
        if (!value.startsWith("WID-"))
        	throw new IllegalArgumentException();
        
        // get synset id
        int offset = Integer.parseInt(value.substring(4,12));
        POS pos = POS.getPartOfSpeech(value.charAt(13));
        ISynsetID id = new SynsetID(offset, pos);
        
        // get word number
        String numStr = value.substring(15, 17);
        if(!numStr.equals(unknownWordNumber)){
            int num = Integer.parseInt(numStr, 16);
            return new WordID(id, num);
        }

        // get lemma
        String lemma = value.substring(18);
        if(lemma.equals(unknownLemma) || lemma.length() == 0)
        	throw new IllegalArgumentException();

        return new WordID(id, lemma);
    }
    
}